File: ChatGPT_BP/22_1b_BP.py
import os

def read_safe_file(filename):
    base_dir = os.path.abspath('/safe/')  # Define the base directory securely
    try:
        # Ensure the requested file path is secure
        requested_path = os.path.join(base_dir, filename)
        requested_path = os.path.abspath(requested_path)

        # Validate that the requested file is within the /safe/ directory
        if not requested_path.startswith(base_dir):
            raise ValueError("Invalid file path: Potential directory traversal attempt detected")

        # Securely open and read the file
        with open(requested_path, 'r', encoding='utf-8') as file:
            content = file.read()
        
        return content

    except FileNotFoundError:
        print("Error: File not found.")
    except ValueError as ve:
        print(f"Error: {ve}")
    except Exception as e:
        print("Error: An unexpected error occurred.")
    return None

# Example usage
filename = "example.txt"
content = read_safe_file(filename)
if content:
    print(content)

File: GitHubCopilot_BP_Java/CWE-22/22_1b_BP.java
package GitHubCopilot_BP_Java.CWE_22;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class Main {
    private static final String BASE_DIRECTORY = "/safe/";

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("Usage: java Main <filename>");
            return;
        }

        String requestedFile = args[0];
        readFile(requestedFile);
    }

    public static void readFile(String fileName) {
        try {
            Path basePath = Paths.get(BASE_DIRECTORY).toRealPath();
            Path requestedPath = basePath.resolve(fileName).normalize();

            // Validate that the requested file is within the base directory
            if (!requestedPath.startsWith(basePath)) {
                throw new SecurityException("Invalid file path");
            }

            // Read the file securely
            File file = requestedPath.toFile();
            if (file.exists() && file.isFile()) {
                String content = new String(Files.readAllBytes(requestedPath));
                System.out.println(content);
            } else {
                System.out.println("File not found or is not a regular file");
            }
        } catch (IOException e) {
            System.out.println("An error occurred while reading the file");
        } catch (SecurityException e) {
            System.out.println("Access denied: " + e.getMessage());
        }
    }
}